<html>
<head>
<meta name="order" content="1" />
<title>WormBase annotation</title>
</head>

<body>

<h1>WormBase annotation</h1>

<h2>Manual curation of gene models in WormBase</h2>

<p>WormBase untertakes targeted manual curation of the gene models in <i>Caenorhabditis elegans</i>. 
Most of the initial <i>C. elegans</i> coding gene structures were originally determined in 1998 
by the <i>ab initio</i> gene prediction program Genefinder (Green, P., unpublished data). 
Curators have spent a number of years editing and improving these inital gene models, using 
supporting data from large-scale transcriptome projects such as Yugi Koharaâ€™s EST libraries 
(Kohara,Y., unpublished data) and more recently, many RNA-Seq projects, including modENCODE. 
Genes with low levels of expression present particular problems because they typically have 
sparse transcript data to support the predicted structure. In these cases, curators rely heavily 
on paralogous and orthologous gene structures (via protein alignments), or by in-depth inspection 
of aligned, conserved genomic regions.</p>

<p>Historically, targets for curation were selected using an automated system that looks for 
discrepencies between the current gene models and the supporting transcriptomic evidence. More 
recently, WormBase has embarked on a theme-based curation strategy, with groups of targets being 
selected according to a particular theme of interest, e.g. members of a specific family, or genes 
involved in a specific pathway.</p>

<p>WormBase also curates gene models in seven additional species (<i>C. briggsae</i>, 
<i>C. brenneri</i>, <i>C. japonica</i>, <i>C. remanei</i>, <i>Onchocerca volvulus</i>, 
<i>Pristionchus pacificus</i> and <i>Brugia malayi</i>). This is mainly driven by 
<ol style="list-style-type:lower-latin">
<li>the curation of a specific group of genes in <i>C. elegans</i> as part of a curation 
theme (see above)</li> 
<li>the availabilty of high-profile data sets (for example a new version of the reference genome, 
or a new large-scale transcriptomics study); and</li>
<li>direct user requests</li>
</ol>

<p>For other species not part of the above core set, WormBase imports predicted gene models 
submitted by the original genome project for that species. No manual curation is performed 
on these gene models.</p>

<h2>WormBase genomes and gene models in Ensembl Metazoa</h2>

<p>The eight core species above, together with selected other species, have their genome and 
gene set represented in Ensembl Metazoa. Data for a species is always drawn from a specific, 
named release of WormBase. Updates are perfomed principally according to demand, with the 
highest demand being for <i>C. elegans</i> (which is updated in Ensembl Metazoa every fifth 
WormBase release).</p>

</body>
</html>
